url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/Receiver_multicast.java

```diff
@@ -0,0 +1,532 @@
+import java.io.ByteArrayInputStream;
+import java.net.DatagramPacket;
+import java.net.InetAddress;
+import java.net.MulticastSocket;
+import java.net.SocketException;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.net.UnknownHostException;
+import java.nio.file.Files;
+import java.util.Random;
+
+import java.io.ObjectInput;
+import java.io.ObjectInputStream;
+import java.util.ConcurrentModificationException;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import RQLibrary.Encoder;
+import RQLibrary.EncodingPacket;
+import RQLibrary.EncodingSymbol;
+import RQLibrary.Partition;
+import RQLibrary.SingularMatrixException;
+import RQLibrary.SourceBlock;
+
+public class Receiver_multicast {
+    
+    // Simulate a lost of the canal
+    public static boolean packetIsLost(float percentageLoss) {
+	Random randomGenerator = new Random();
+	int rand = randomGenerator.nextInt(100);
+	if ( rand < (int) percentageLoss ) {
+	    return true;
+	}
+	return false;
+	
+    }
+    
+    public static final int lookForTheAnswer(InetAddress groupeIP, int port) throws IOException {
+	int packetData = 0;
+	MulticastSocket socketReception;
+	socketReception = new MulticastSocket(port);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *socketReception*. The resource is referenced at lines 44, 49. The resource is closed at line 58. There are other execution paths that do not contain closure statements, for example, when MulticastSocket.joinGroup throws an exception. Ensure *socketReception* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).